[BUG] Go generator throws java.lang.StringIndexOutOfBoundsException generating example code for recursive model

#### Bug Report Checklist

- [X] Have you provided a full/minimal spec to reproduce the issue?
- [X] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [X] Have you [tested with the latest master](https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-to-test-with-the-latest-master-of-openapi-generator) to confirm the issue still exists?
- [X] Have you searched for related issues/PRs?
  - The problem seems specific to the Go generator. There are similar issues for other languages but not this one.
- [X] What's the actual output vs expected output?
  - No exception is thrown 
- [ ] [Optional] Sponsorship to speed up the bug fix or feature request ([example](https://github.com/OpenAPITools/openapi-generator/issues/6178))

<!--
Please follow the issue template below for bug reports.
Also please indicate in the issue title which language/library is concerned. Eg:  [BUG][JAVA] Bug generating foo with bar 
-->

##### Description

The `go` generator has a bug that causes API generation to fail while executing `contrusctExampleCode`. As far as I could check, the problem is triggered when the following conditions are met:
 - The offending property is part of a request body.
 - The offending property is marked as required (so example code is generated).
 - The offending property is nested and references one of its parents, hence using recursion.

Please see below a minimal example spec that reproduces the issue.

##### openapi-generator version

`6.6.0` through `openapitools/openapi-generator-cli` NPM package, but also reproduced by building it locally.

##### OpenAPI declaration file content or url

```yaml
openapi: 3.0.3
info:
  title: "Small reproducer for go generator bug"
  version: 0.0.1
paths:
  /sample:
    post:
      operationId: getSample
      requestBody:
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/Sample'
        required: true
      responses:
        '204':
          description: successful operation
components:
  schemas:
    Sample:
      type: object
      required:
        - predicate
      properties:
        predicate:
          items:
            $ref: '#/components/schemas/Predicate'
    Predicate:
      oneOf:
        - $ref: '#/components/schemas/PredicateAnd'
    PredicateAnd:
      type: object
      required:
        - and
      properties:
        and:
          type: array
          items:
            $ref: '#/components/schemas/Predicate'
```

##### Generation Details

Full exception:
```
Exception in thread "main" java.lang.RuntimeException: Could not generate api file for 'Default'
        at org.openapitools.codegen.DefaultGenerator.generateApis(DefaultGenerator.java:701)
        at org.openapitools.codegen.DefaultGenerator.generate(DefaultGenerator.java:921)
        at org.openapitools.codegen.cmd.Generate.execute(Generate.java:473)
        at org.openapitools.codegen.cmd.OpenApiGeneratorCommand.run(OpenApiGeneratorCommand.java:32)
        at org.openapitools.codegen.OpenAPIGenerator.main(OpenAPIGenerator.java:66)
Caused by: java.lang.StringIndexOutOfBoundsException: begin 1, end 0, length 0
        at java.base/java.lang.String.checkBoundsBeginEnd(String.java:4604)
        at java.base/java.lang.String.substring(String.java:2707)
        at java.base/java.lang.String.substring(String.java:2680)
        at org.openapitools.codegen.languages.GoClientCodegen.constructExampleCode(GoClientCodegen.java:677)
        at org.openapitools.codegen.languages.GoClientCodegen.constructExampleCode(GoClientCodegen.java:629)
        at org.openapitools.codegen.languages.GoClientCodegen.constructExampleCode(GoClientCodegen.java:585)
        at org.openapitools.codegen.languages.GoClientCodegen.constructExampleCode(GoClientCodegen.java:688)
        at org.openapitools.codegen.languages.GoClientCodegen.constructExampleCode(GoClientCodegen.java:677)
        at org.openapitools.codegen.languages.GoClientCodegen.constructExampleCode(GoClientCodegen.java:629)
        at org.openapitools.codegen.languages.GoClientCodegen.constructExampleCode(GoClientCodegen.java:585)
        at org.openapitools.codegen.languages.GoClientCodegen.constructExampleCode(GoClientCodegen.java:688)
        at org.openapitools.codegen.languages.GoClientCodegen.constructExampleCode(GoClientCodegen.java:677)
        at org.openapitools.codegen.languages.GoClientCodegen.constructExampleCode(GoClientCodegen.java:629)
        at org.openapitools.codegen.languages.GoClientCodegen.constructExampleCode(GoClientCodegen.java:585)
        at org.openapitools.codegen.languages.GoClientCodegen.constructExampleCode(GoClientCodegen.java:688)
        at org.openapitools.codegen.languages.GoClientCodegen.constructExampleCode(GoClientCodegen.java:555)
        at org.openapitools.codegen.languages.GoClientCodegen.postProcessOperationsWithModels(GoClientCodegen.java:487)
        at org.openapitools.codegen.DefaultGenerator.processOperations(DefaultGenerator.java:1243)
        at org.openapitools.codegen.DefaultGenerator.generateApis(DefaultGenerator.java:605)
        ... 4 more
```

##### Steps to reproduce

```shell
npx @openapitools/openapi-generator-cli generate -i /path/to/spec.yaml -g go
```

##### Related issues/PRs

 - #2791

##### Suggest a fix

As far as I could see, there is code to prevent infinite recursion, but at some point it returns an empty string to stop it, which later causes the bug since [line 697](https://github.com/OpenAPITools/openapi-generator/blob/f0d439fbce49d11112faf8044bfa7380945d601e/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/GoClientCodegen.java#LL697C1-L697C125) executes `.substring(1)` over this empty line.

This operation seems to be intended to remove the `*` prefix used by other code branches, so a good alternative could be making that explicit instead of directly assuming the string has contents.

--- a/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/GoClientCodegen.java
+++ b/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/GoClientCodegen.java
@@ -694,7 +694,7 @@ public class GoClientCodegen extends AbstractGoCodegen {
             if (modelMaps.get(subModel) == null) {
                 oneOf = "new(" + subModel + ")";// a primitive type
             } else {
-                oneOf = constructExampleCode(modelMaps.get(subModel), modelMaps, processedModelMap, depth + 1).substring(1);
+                oneOf = StringUtils.removeStart(constructExampleCode(modelMaps.get(subModel), modelMaps, processedModelMap, depth + 1), "*");
             }
             return goImportAlias + "." + model + "{" + typeToName(subModel) + ": " + oneOf + "}";
         } else {

